CFE-2989: Human readable JSON5 dumping of LMDB databases

olehermanse opened the pull request:
cf-check can use our knowledge of what these databases contain (structs), as well as the JSON code in libntech, to print the information in a readable way. Here's a comparison:

**mdb_dump:**

```
root@dev core $ mdb_dump -np /var/cfengine/state/cf_lastseen.lmdb
VERSION=3
format=print
type=btree
mapsize=104857600
maxreaders=126
db_pagesize=4096
HEADER=END
 a192.168.100.10\00
 MD5=0924931c0116d2640e609bfbb424dc18\00
 kMD5=0924931c0116d2640e609bfbb424dc18\00
 192.168.100.10\00
 qoMD5=0924931c0116d2640e609bfbb424dc18\00
 \cd\96z]\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00
DATA=END
root@dev core $
```

**lmdump:**

```
root@dev core $ lmdump -a /var/cfengine/state/cf_lastseen.lmdb
key: 0x7f3bc38abf34[16] a192.168.100.10, data: 0x7f3bc38abf44[37] MD5=0924931c0116d2640e609bfbb424dc18
key: 0x7f3bc38abf72[38] kMD5=0924931c0116d2640e609bfbb424dc18, data: 0x7f3bc38abf98[15] 192.168.100.10
key: 0x7f3bc38abfb0[39] qoMD5=0924931c0116d2640e609bfbb424dc18, data: 0x7f3bc38abfd7[40] ͖z]
root@dev core $
```

**cf-check dump:**

```
root@dev core $ cf-check dump /var/cfengine/state/cf_lastseen.lmdb
{
        "a192.168.100.10": "MD5=0924931c0116d2640e609bfbb424dc18",
        "kMD5=0924931c0116d2640e609bfbb424dc18": "192.168.100.10",
        "qoMD5=0924931c0116d2640e609bfbb424dc18": {"Q":{"dq":0.0000,"expect":0.0000,"q":0.0000,"var":0.0000},"lastseen":1568315085},
}
root@dev core $
```

**Using json5 and jq to pretty-print "normal" JSON:**

```
root@dev core $ cf-check dump /var/cfengine/state/cf_lastseen.lmdb | json5 | jq
{
  "a192.168.100.10": "MD5=0924931c0116d2640e609bfbb424dc18",
  "kMD5=0924931c0116d2640e609bfbb424dc18": "192.168.100.10",
  "qoMD5=0924931c0116d2640e609bfbb424dc18": {
    "Q": {
      "dq": 0,
      "expect": 0,
      "q": 0,
      "var": 0
    },
    "lastseen": 1568315085
  }
}
root@dev core $
```

olehermanse commented:
@cf-bottom jenkins

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=2748)](https://ci.cfengine.com//job/pr-pipeline/2748/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2748/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2748/

nickanderson commented:
@cf-bottom jenkins, please

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=2765)](https://ci.cfengine.com//job/pr-pipeline/2765/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2765/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2765/

vpodzime commented on cf-check/dump.c:
```diff
@@ -346,6 +346,31 @@ int dump_db(const char *file, const dump_mode mode)
     return 0;
 }
 
+static bool matches_option(
```
Looks like a good candidate for a function in *libutils*.

olehermanse replied:
Yep, I'll do it in a follow-up PR.

vpodzime commented on cf-check/dump.c:
```diff
@@ -44,14 +44,31 @@ void dump_print_json_string(const char *const data, const size_t size)
+
+    // Unfortunately, the packages_installed_apt_get.lmdb database
+    // needs some exceptions:
+    assert((size == 1) || data[size -1] == '\n' || (length == (size - 1)));
```
Missing space in `size -1`.

olehermanse replied:
Fixed in a later commit.

vpodzime commented on cf-check/dump.c:
```diff
@@ -438,25 +439,17 @@ static bool matches_option(
     const char *const *filenames = argv + 1;
     size_t filenames_len = argc - 1;
-    if (argv[1][0] == '-')
+    if (filenames_len > 0 && filenames[0] != NULL && filenames[0][0] == '-')
```
`filenames[0][0] == '-'` is really confusing. What about just checking `argv` here and define `filenames` and `filenames_len` when it's clear where the file names start?

olehermanse replied:
I considered both, and thought this was cleaner (less thinking about indices / offsets), but then I'll change it.

vpodzime commented on cf-check/dump.c:
```diff
@@ -467,11 +467,18 @@ int dump_main(int argc, const char *const *const argv)
 
     if (filenames_len > 0 && filenames[0] != NULL && filenames[0][0] == '-')
     {
         print_usage();
```
`print_usage()` should list the options and use `|` to mark them as mutually exclusive.

olehermanse replied:
Added.

vpodzime commented on cf-check/dump.c:
```diff
@@ -266,72 +251,100 @@ void nice_print_persistent_class(const MDB_val value)
+    if (!was_printed)
     {
-        dump_print_json_string_nice(value.mv_data, value.mv_size);
+        dump_print_json_string(value.mv_data, value.mv_size, strip_strings);
```
I'd prefer `dump_print_json_string(...)` in the `else` branch of the long `if-else if-else` above and in `else` of the `if (structs)`. This `was_printed` is a bit confusing to me.

olehermanse replied:
ACK, fixed.

vpodzime commented on cf-check/dump.c:
```diff
@@ -266,72 +251,100 @@ void nice_print_persistent_class(const MDB_val value)
 {
     assert(file != NULL);
 
+    const bool strip_strings = ((mode == DUMP_MODE_NICE) ? true : false);
```
Why not just `strip_strings = (mode == DUMP_MODE_NICE)`?

olehermanse replied:
Haha, too much copy pasting and refactoring. Fixed, thanks!

vpodzime commented on cf-check/dump.c:
```diff
@@ -266,72 +251,100 @@ void nice_print_persistent_class(const MDB_val value)
 
+    const bool strip_strings = ((mode == DUMP_MODE_NICE) ? true : false);
+    const bool structs =
+        ((mode == DUMP_MODE_NICE || mode == DUMP_MODE_PORTABLE) ? true : false);
```
And similar here.

olehermanse replied:
Thanks!

olehermanse commented:
@vpodzime Addressed your comment in a last separate commit. In this case, I personally prefer to keep the commit history as this, but tell me if you want to squash something. (Doing a full rebase and working your comments into the correct commits is a decent amount of work, and also makes it more difficult to re-review.)

olehermanse commented:
@cf-bottom jenkins

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=2772)](https://ci.cfengine.com//job/pr-pipeline/2772/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2772/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2772/

craigcomstock commented on cf-check/db_structs.h:
```diff
@@ -0,0 +1,152 @@
+//     double expect;
+//     double var;
+//     double dq;
+// } QPoint;
```
Why is this one commented out?

olehermanse replied:
Because it's in statistics.h , which is in libutils and can be included.

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=2773)](https://ci.cfengine.com//job/pr-pipeline/2773/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2773/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2773/

olehermanse commented:
Cherry-pick: https://github.com/cfengine/core/pull/3845
